// @zed-industries/agent-client-protocol/dist/schema.js
export const AGENT_METHODS = {
    authenticate: "authenticate",
    initialize: "initialize",
    session_cancel: "session/cancel",
    session_load: "session/load",
    session_new: "session/new",
    session_prompt: "session/prompt",
};
export const CLIENT_METHODS = {
    fs_read_text_file: "fs/read_text_file",
    fs_write_text_file: "fs/write_text_file",
    session_request_permission: "session/request_permission",
    session_update: "session/update",
    terminal_create: "terminal/create",
    terminal_kill: "terminal/kill",
    terminal_output: "terminal/output",
    terminal_release: "terminal/release",
    terminal_wait_for_exit: "terminal/wait_for_exit",
};
export const PROTOCOL_VERSION = 1;
import { z } from "zod";
/** @internal */
export const writeTextFileRequestSchema = z.object({
    content: z.string(),
    path: z.string(),
    sessionId: z.string(),
});
/** @internal */
export const readTextFileRequestSchema = z.object({
    limit: z.number().optional().nullable(),
    line: z.number().optional().nullable(),
    path: z.string(),
    sessionId: z.string(),
});
/** @internal */
export const roleSchema = z.union([z.literal("assistant"), z.literal("user")]);
/** @internal */
export const textResourceContentsSchema = z.object({
    mimeType: z.string().optional().nullable(),
    text: z.string(),
    uri: z.string(),
});
/** @internal */
export const blobResourceContentsSchema = z.object({
    blob: z.string(),
    mimeType: z.string().optional().nullable(),
    uri: z.string(),
});
/** @internal */
export const toolKindSchema = z.union([
    z.literal("read"),
    z.literal("edit"),
    z.literal("delete"),
    z.literal("move"),
    z.literal("search"),
    z.literal("execute"),
    z.literal("think"),
    z.literal("fetch"),
    z.literal("other"),
]);
/** @internal */
export const toolCallStatusSchema = z.union([
    z.literal("pending"),
    z.literal("in_progress"),
    z.literal("completed"),
    z.literal("failed"),
]);
/** @internal */
export const sessionIdSchema = z.string();
/** @internal */
export const writeTextFileResponseSchema = z.null();
/** @internal */
export const readTextFileResponseSchema = z.object({
    content: z.string(),
});
/** @internal */
export const requestPermissionResponseSchema = z.object({
    outcome: z.union([
        z.object({
            outcome: z.literal("cancelled"),
        }),
        z.object({
            optionId: z.string(),
            outcome: z.literal("selected"),
        }),
    ]),
});
/** @internal */
export const createTerminalResponseSchema = z.object({
    terminalId: z.string(),
});
/** @internal */
export const releaseTerminalResponseSchema = z.null();
/** @internal */
export const waitForTerminalExitResponseSchema = z.object({
    exitCode: z.number().optional().nullable(),
    signal: z.string().optional().nullable(),
});
/** @internal */
export const killTerminalResponseSchema = z.null();
/** @internal */
export const cancelNotificationSchema = z.object({
    sessionId: z.string(),
});
/** @internal */
export const authenticateRequestSchema = z.object({
    methodId: z.string(),
});
/** @internal */
export const annotationsSchema = z.object({
    audience: z.array(roleSchema).optional().nullable(),
    lastModified: z.string().optional().nullable(),
    priority: z.number().optional().nullable(),
});
/** @internal */
export const embeddedResourceResourceSchema = z.union([
    textResourceContentsSchema,
    blobResourceContentsSchema,
]);
/** @internal */
export const authenticateResponseSchema = z.null();
/** @internal */
export const loadSessionResponseSchema = z.null();
/** @internal */
export const promptResponseSchema = z.object({
    stopReason: z.union([
        z.literal("end_turn"),
        z.literal("max_tokens"),
        z.literal("max_turn_requests"),
        z.literal("refusal"),
        z.literal("cancelled"),
    ]),
});
/** @internal */
export const unstructuredCommandInputSchema = z.object({
    hint: z.string(),
});
/** @internal */
export const permissionOptionSchema = z.object({
    kind: z.union([
        z.literal("allow_once"),
        z.literal("allow_always"),
        z.literal("reject_once"),
        z.literal("reject_always"),
    ]),
    name: z.string(),
    optionId: z.string(),
});
/** @internal */
export const toolCallContentSchema = z.union([
    z.object({
        content: z.union([
            z.object({
                annotations: annotationsSchema.optional().nullable(),
                text: z.string(),
                type: z.literal("text"),
            }),
            z.object({
                annotations: annotationsSchema.optional().nullable(),
                data: z.string(),
                mimeType: z.string(),
                type: z.literal("image"),
                uri: z.string().optional().nullable(),
            }),
            z.object({
                annotations: annotationsSchema.optional().nullable(),
                data: z.string(),
                mimeType: z.string(),
                type: z.literal("audio"),
            }),
            z.object({
                annotations: annotationsSchema.optional().nullable(),
                description: z.string().optional().nullable(),
                mimeType: z.string().optional().nullable(),
                name: z.string(),
                size: z.number().optional().nullable(),
                title: z.string().optional().nullable(),
                type: z.literal("resource_link"),
                uri: z.string(),
            }),
            z.object({
                annotations: annotationsSchema.optional().nullable(),
                resource: embeddedResourceResourceSchema,
                type: z.literal("resource"),
            }),
        ]),
        type: z.literal("content"),
    }),
    z.object({
        newText: z.string(),
        oldText: z.string().optional().nullable(),
        path: z.string(),
        type: z.literal("diff"),
    }),
    z.object({
        terminalId: z.string(),
        type: z.literal("terminal"),
    }),
]);
/** @internal */
export const toolCallLocationSchema = z.object({
    line: z.number().optional().nullable(),
    path: z.string(),
});
/** @internal */
export const envVariableSchema = z.object({
    name: z.string(),
    value: z.string(),
});
/** @internal */
export const terminalOutputRequestSchema = z.object({
    sessionId: sessionIdSchema,
    terminalId: z.string(),
});
/** @internal */
export const releaseTerminalRequestSchema = z.object({
    sessionId: sessionIdSchema,
    terminalId: z.string(),
});
/** @internal */
export const waitForTerminalExitRequestSchema = z.object({
    sessionId: sessionIdSchema,
    terminalId: z.string(),
});
/** @internal */
export const killTerminalRequestSchema = z.object({
    sessionId: sessionIdSchema,
    terminalId: z.string(),
});
/** @internal */
export const terminalExitStatusSchema = z.object({
    exitCode: z.number().optional().nullable(),
    signal: z.string().optional().nullable(),
});
/** @internal */
export const fileSystemCapabilitySchema = z.object({
    readTextFile: z.boolean().optional(),
    writeTextFile: z.boolean().optional(),
});
/** @internal */
export const mcpServerSchema = z.object({
    args: z.array(z.string()),
    command: z.string(),
    env: z.array(envVariableSchema),
    name: z.string(),
});
/** @internal */
export const loadSessionRequestSchema = z.object({
    cwd: z.string(),
    mcpServers: z.array(mcpServerSchema),
    sessionId: z.string(),
});
/** @internal */
export const contentBlockSchema = z.union([
    z.object({
        annotations: annotationsSchema.optional().nullable(),
        text: z.string(),
        type: z.literal("text"),
    }),
    z.object({
        annotations: annotationsSchema.optional().nullable(),
        data: z.string(),
        mimeType: z.string(),
        type: z.literal("image"),
        uri: z.string().optional().nullable(),
    }),
    z.object({
        annotations: annotationsSchema.optional().nullable(),
        data: z.string(),
        mimeType: z.string(),
        type: z.literal("audio"),
    }),
    z.object({
        annotations: annotationsSchema.optional().nullable(),
        description: z.string().optional().nullable(),
        mimeType: z.string().optional().nullable(),
        name: z.string(),
        size: z.number().optional().nullable(),
        title: z.string().optional().nullable(),
        type: z.literal("resource_link"),
        uri: z.string(),
    }),
    z.object({
        annotations: annotationsSchema.optional().nullable(),
        resource: embeddedResourceResourceSchema,
        type: z.literal("resource"),
    }),
]);
/** @internal */
export const authMethodSchema = z.object({
    description: z.string().optional().nullable(),
    id: z.string(),
    name: z.string(),
});
/** @internal */
export const promptCapabilitiesSchema = z.object({
    audio: z.boolean().optional(),
    embeddedContext: z.boolean().optional(),
    image: z.boolean().optional(),
});
/** @internal */
export const availableCommandInputSchema = unstructuredCommandInputSchema;
/** @internal */
export const planEntrySchema = z.object({
    content: z.string(),
    priority: z.union([z.literal("high"), z.literal("medium"), z.literal("low")]),
    status: z.union([
        z.literal("pending"),
        z.literal("in_progress"),
        z.literal("completed"),
    ]),
});
/** @internal */
export const clientNotificationSchema = cancelNotificationSchema;
/** @internal */
export const createTerminalRequestSchema = z.object({
    args: z.array(z.string()).optional(),
    command: z.string(),
    cwd: z.string().optional().nullable(),
    env: z.array(envVariableSchema).optional(),
    outputByteLimit: z.number().optional().nullable(),
    sessionId: sessionIdSchema,
});
/** @internal */
export const terminalOutputResponseSchema = z.object({
    exitStatus: terminalExitStatusSchema.optional().nullable(),
    output: z.string(),
    truncated: z.boolean(),
});
/** @internal */
export const newSessionRequestSchema = z.object({
    cwd: z.string(),
    mcpServers: z.array(mcpServerSchema),
});
/** @internal */
export const promptRequestSchema = z.object({
    prompt: z.array(contentBlockSchema),
    sessionId: z.string(),
});
/** @internal */
export const sessionNotificationSchema = z.object({
    sessionId: z.string(),
    update: z.union([
        z.object({
            content: contentBlockSchema,
            sessionUpdate: z.literal("user_message_chunk"),
        }),
        z.object({
            content: contentBlockSchema,
            sessionUpdate: z.literal("agent_message_chunk"),
        }),
        z.object({
            content: contentBlockSchema,
            sessionUpdate: z.literal("agent_thought_chunk"),
        }),
        z.object({
            content: z.array(toolCallContentSchema).optional(),
            kind: z
                .union([
                z.literal("read"),
                z.literal("edit"),
                z.literal("delete"),
                z.literal("move"),
                z.literal("search"),
                z.literal("execute"),
                z.literal("think"),
                z.literal("fetch"),
                z.literal("other"),
            ])
                .optional(),
            locations: z.array(toolCallLocationSchema).optional(),
            rawInput: z.record(z.unknown()).optional(),
            rawOutput: z.record(z.unknown()).optional(),
            sessionUpdate: z.literal("tool_call"),
            status: z
                .union([
                z.literal("pending"),
                z.literal("in_progress"),
                z.literal("completed"),
                z.literal("failed"),
            ])
                .optional(),
            title: z.string(),
            toolCallId: z.string(),
        }),
        z.object({
            content: z.array(toolCallContentSchema).optional().nullable(),
            kind: toolKindSchema.optional().nullable(),
            locations: z.array(toolCallLocationSchema).optional().nullable(),
            rawInput: z.record(z.unknown()).optional(),
            rawOutput: z.record(z.unknown()).optional(),
            sessionUpdate: z.literal("tool_call_update"),
            status: toolCallStatusSchema.optional().nullable(),
            title: z.string().optional().nullable(),
            toolCallId: z.string(),
        }),
        z.object({
            entries: z.array(planEntrySchema),
            sessionUpdate: z.literal("plan"),
        }),
    ]),
});
/** @internal */
export const toolCallUpdateSchema = z.object({
    content: z.array(toolCallContentSchema).optional().nullable(),
    kind: toolKindSchema.optional().nullable(),
    locations: z.array(toolCallLocationSchema).optional().nullable(),
    rawInput: z.record(z.unknown()).optional(),
    rawOutput: z.record(z.unknown()).optional(),
    status: toolCallStatusSchema.optional().nullable(),
    title: z.string().optional().nullable(),
    toolCallId: z.string(),
});
/** @internal */
export const clientCapabilitiesSchema = z.object({
    fs: fileSystemCapabilitySchema.optional(),
    terminal: z.boolean().optional(),
});
/** @internal */
export const agentCapabilitiesSchema = z.object({
    loadSession: z.boolean().optional(),
    promptCapabilities: promptCapabilitiesSchema.optional(),
});
/** @internal */
export const availableCommandSchema = z.object({
    description: z.string(),
    input: availableCommandInputSchema.optional().nullable(),
    name: z.string(),
});
/** @internal */
export const clientResponseSchema = z.union([
    writeTextFileResponseSchema,
    readTextFileResponseSchema,
    requestPermissionResponseSchema,
    createTerminalResponseSchema,
    terminalOutputResponseSchema,
    releaseTerminalResponseSchema,
    waitForTerminalExitResponseSchema,
    killTerminalResponseSchema,
]);
/** @internal */
export const agentNotificationSchema = sessionNotificationSchema;
/** @internal */
export const requestPermissionRequestSchema = z.object({
    options: z.array(permissionOptionSchema),
    sessionId: z.string(),
    toolCall: toolCallUpdateSchema,
});
/** @internal */
export const initializeRequestSchema = z.object({
    clientCapabilities: clientCapabilitiesSchema.optional(),
    protocolVersion: z.number(),
});
/** @internal */
export const initializeResponseSchema = z.object({
    agentCapabilities: agentCapabilitiesSchema.optional(),
    authMethods: z.array(authMethodSchema).optional(),
    protocolVersion: z.number(),
});
/** @internal */
export const newSessionResponseSchema = z.object({
    availableCommands: z.array(availableCommandSchema).optional(),
    sessionId: z.string(),
});
/** @internal */
export const clientRequestSchema = z.union([
    writeTextFileRequestSchema,
    readTextFileRequestSchema,
    requestPermissionRequestSchema,
    createTerminalRequestSchema,
    terminalOutputRequestSchema,
    releaseTerminalRequestSchema,
    waitForTerminalExitRequestSchema,
    killTerminalRequestSchema,
]);
/** @internal */
export const agentRequestSchema = z.union([
    initializeRequestSchema,
    authenticateRequestSchema,
    newSessionRequestSchema,
    loadSessionRequestSchema,
    promptRequestSchema,
]);
/** @internal */
export const agentResponseSchema = z.union([
    initializeResponseSchema,
    authenticateResponseSchema,
    newSessionResponseSchema,
    loadSessionResponseSchema,
    promptResponseSchema,
]);
/** @internal */
export const agentClientProtocolSchema = z.union([
    clientRequestSchema,
    clientResponseSchema,
    clientNotificationSchema,
    agentRequestSchema,
    agentResponseSchema,
    agentNotificationSchema,
]);
